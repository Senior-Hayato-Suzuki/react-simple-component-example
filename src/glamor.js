import React from "react";
import cn from "classnames";
import {css} from "glamor";

import createSimple from "./core";

const defaultNamePat = /^simple\(.*\)/;
var counter = 0;

const simple = createSimple(css, (
    self,
    Component,
    rule,
    {className, ...otherProps},
    altStyleName,
) => {
    const num = counter++;

    const makeDebugClass = className =>
        className && !defaultNamePat.test(className)
            ? className + "___" + num
            : null;

    const props = {
        ...otherProps,
        className: cn(
            makeDebugClass(self.displayName),
            makeDebugClass(altStyleName),
            className,
            String(rule),
        ),
    };

    return <Component {...props} />;
});

// This is experimental API. Do not use!
// A div which mirrors the View component in React Native
const viewStyles = {
    boxSizing: "border-box",
    position: "relative",
    display: "flex",
    flexDirection: "column",
    alignItems: "stretch",
    flexShrink: 0,
    alignContent: "flex-start",
    minWidth: 0,
};
export const View = simple("div", viewStyles);
View.displayName = "View";

export {css};
export default simple;
