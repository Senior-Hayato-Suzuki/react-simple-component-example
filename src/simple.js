import React from "react";
import cn from "classnames";
import {css} from "glamor";
import omit from "lodash/fp/omit";
import pick from "lodash/fp/pick";
import pickBy from "lodash/fp/pickBy";
import deepmerge from "deepmerge";

const pickTruthty = pickBy(Boolean);

function getDisplayName(Component) {
    if (typeof Component === "string") {
        return Component;
    }

    return Component.displayName || Component.name || "Anonymous";
}

var compnentCounter = 0;

function simple(Component, styles, alts={}, options={}) {
    const num = compnentCounter++;
    const makeDebugClass = className => className
            ? className + "___" + num
            : null;

    const level = options._level || 0;

    if (Component._styleWrapped) {
        return simple(
            Component._styleWrapped.Component,
            deepmerge(Component._styleWrapped.styles, styles),
            deepmerge(Component._styleWrapped.alts, alts),
            {
                _level: level + 1,
            }
        );
    }

    const rules = {__base: css(styles)};

    const altProps = Object.keys(alts);


    for (let key in alts) {
        rules[key] = css(deepmerge(styles, alts[key]));
    }

    function StyleWrapped({className, ...otherProps}) {

        const passProps = omit(altProps, otherProps);
        const alt = Object.keys(pickTruthty(pick(altProps, otherProps)));


        if (alt.length > 1) {
            throw new Error("Too many alt props: " + alt.join(", "));
        }

        const rule = rules[alt[0] || "__base"];


        const props = {
            ...passProps,
            className: cn(makeDebugClass(StyleWrapped.displayName), makeDebugClass(alt[0]), className, String(rule)),
        };

        return <Component {...props} />;
    }

    StyleWrapped._styleWrapped = {
        Component,
        styles,
        alts,
    };

    StyleWrapped.displayName = "simple(" + getDisplayName(Component) + ")";

    StyleWrapped.create = el => simple(el, styles, alts);


    return StyleWrapped;
}


const viewStyles = {
    boxSizing: "border-box",
    position: "relative",
    display: "flex",
    flexDirection: "column",
    alignItems: "stretch",
    flexShrink: 0,
    alignContent: "flex-start",
    border: 0,
    margin: 0,
    padding: 0,
    minWidth: 0,
};

export const View = simple("div", viewStyles);
View.displayName = "View";


export {css};
export default simple;
