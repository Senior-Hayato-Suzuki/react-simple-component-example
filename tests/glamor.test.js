import React from "react";
import simple from "../src/glamor";
import {css} from "glamor";
import renderer from "react-test-renderer";

// jest.mock("glamor", () => ({
//     css: jest.fn(() => "glamor-generated-class-name"),
// }));

describe("glamor backend", () => {
    // beforeEach(() => {
    //     css.mockClear();
    // });

    test("renders class name custom classname", () => {
        const Button = simple("button", {
            padding: 10,
        });

        const tree = renderer
            .create(<Button className="custom-class-name" />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
