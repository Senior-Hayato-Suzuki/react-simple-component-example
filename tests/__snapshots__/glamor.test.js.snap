exports[`glamor backend renders class name 1`] = `
<button
  className="Button___1 glamor-generated-class-name" />
`;

exports[`glamor backend renders the alt class 1`] = `
<button
  className="Button___4 red___4 glamor-generated-class-name" />
`;
